### PR TITLE
Adding more vector calls for -fveclib=AMDLIBM

### PR DESCRIPTION
**Pull request by rohitaggarwal007**
Hi,

AMD has it's own implementation of vector calls. New vector calls are introduced in the library.
Please refer [https://github.com/amd/aocl-libm-ose]

### CODE REVIEW

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll [hunk 1]
```diff
@@ -1600,4 +1653,163 @@ for.end:
   ret void
 }
 
+define void @exp10_f64(ptr nocapture %varray) {
+; CHECK-LABEL: @exp10_f64(
+; CHECK-VF2:    [[TMP5:%.*]] = call <2 x double> @amd_vrd2_exp10(<2 x double> [[TMP4:%.*]])
+; CHECK-VF4-NOT:    call <4 x double> @amd_vrd4_exp10(<4 x double> [[TMP4:%.*]])
```
Comment (RKSimon) [inline llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1659]: Check for @llvm.exp10.v4f64 instead (etc.)? That's what we do in the rest of the file.

Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1659]: there are no such intrinsic present for exp10. I tried but it is not emitting llvm.exp10.v4f64. Instead, it generating @emp10.
image

Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1659]: @RKSimon Please check the above comment

Comment (RKSimon) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1659]: Are we just missing exp10 handling in isTriviallyVectorizable?

Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1659]: Yes, case for exp10 is not handled in bool llvm::isTriviallyVectorizable(Intrinsic::ID ID)

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll [hunk 2]
```diff
@@ -1600,4 +1653,163 @@ for.end:
...
+  %exitcond = icmp eq i64 %iv.next, 1000
+  br i1 %exitcond, label %for.end, label %for.body
+
+for.end:
+  ret void
+}
+
+define void @exp10_f64_intrinsic(ptr nocapture %varray) {
+; CHECK-LABEL: @exp10_f64_intrinsic(
+; CHECK-VF2:    [[TMP5:%.*]] = call <2 x double> @amd_vrd2_exp10(<2 x double> [[TMP4:%.*]])
```
Comment (RKSimon) [inline llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1708]: Missing VF2 entries

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll [hunk 3]
```diff
@@ -1600,4 +1653,163 @@ for.end:
...
+  br i1 %exitcond, label %for.end, label %for.body
+
+for.end:
+  ret void
+}
+
+define void @exp10_f32(ptr nocapture %varray) {
+; CHECK-LABEL: @exp10_f32(
+; CHECK-VF4:    [[TMP5:%.*]] = call <4 x float> @amd_vrs4_exp10f(<4 x float> [[TMP4:%.*]])
+; CHECK-VF8-NOT:    [[TMP5:%.*]] = call <8 x float> @amd_vrs8_exp10f(<8 x float> [[TMP4:%.*]])
```
Comment (RKSimon) [inline llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1684]: Its OK if you want to keep the CHECK-VF*-NOT line as well, but I'd expect the CHECK-VF* line (llvm.exp10.v8f32?) etc. as well.

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll [hunk 4]
```diff
@@ -1444,6 +1444,59 @@ for.end:
...
+  %iv.next = add nuw nsw i64 %iv, 1
+  %exitcond = icmp eq i64 %iv.next, 1000
+  br i1 %exitcond, label %for.end, label %for.body
+
+for.end:
+  ret void
+}
+
+
 define void @log10_f32(ptr nocapture %varray) {
```
Comment (RKSimon) [inline llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1500]: (pedantic) we're typically ordering the tests as f64/f32/f64_intrinsic/f32_intrinsic sets

**Review by RKSimon — COMMENTED** [review]
Any luck with adding exp10 to isTriviallyVectorizable?

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll (hunk 3 above)
Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1684]: I will start of add the support of exp10 in isTriviallyVectorizable

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll (hunk 4 above)
Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1500]: Yes, It's help to cover up the cases. Is their some other convention that should be followed?

Comment (RKSimon) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1500]: mainly just consistency in grouping/ordering/naming to make it easier for people to find related tests

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll (hunk 2 above)
Comment (RKSimon) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll:1708]: Sorry I added the comment to the wrong test - I meant that every test should have VF2/4/8/16 test checks

File: llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll (hunk 3 above)
Comment (rohitaggarwal007) [reply llvm/test/Transforms/LoopVectorize/X86/amdlibm-calls.ll]: Now exp10 is generated as llvm.exp10.v8f32 for VF=8 and respectively

**Review by RKSimon — APPROVED** [review]
LGTM

**Review by farzonl — APPROVED** [review]

**Comment by rohitaggarwal007** [comment]
> @nikic Please review the pending changes

Gentle Reminder...

**Comment by RKSimon** [comment]
@rohitaggarwal007 Please can you rebase and then I will commit this

**Comment by rohitaggarwal007** [comment]
> @rohitaggarwal007 Please can you rebase and then I will commit this

Sure, Thanks

**Comment by rohitaggarwal007** [comment]
@RKSimon , I have rebase the code.

**Comment by rohitaggarwal007** [comment]
Thanks @RKSimon

**Pull request merged**